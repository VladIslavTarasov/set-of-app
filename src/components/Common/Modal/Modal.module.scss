.container {
  position: fixed;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  background-color: var(--black2);
}

.modal {
  position: fixed;
  z-index: var(--z-modal);
  top: 50%;
  left: 50%;
  overflow: hidden;
  min-width: 400px;
  max-width: 600px;
  box-sizing: border-box;
  padding: var(--offset-sm, --offset-md);
  background-color: var(--white);
  border-radius: var(--radius-md);
  box-shadow: var(--shadow1);
  transform: translate(-50%, -50%);
}

.header {
  padding: var(--offset-md);
  margin: -16px -16px 0;
  background-color: var(--gray1);
}

.title {
  padding: 0;
  margin: 0;
  font-size: var(--fs-xl);
  font-weight: 600;
}

.content {
  margin: var(--offset-md);
}

.text {
  padding: 0;
  margin: 0;
  font-size: var(--fs-sm);
  font-weight: 400;

  &:not(:first-child) {
    margin-bottom: var(--offset-xs);
  }
}
