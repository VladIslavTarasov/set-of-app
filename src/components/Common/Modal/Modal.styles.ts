import { createUseStyles } from 'react-jss';

import { Theme } from 'theme/types/theme.types';

export const useStyles = createUseStyles(
  {
    container: ({ theme }: { theme: Theme }) => ({
      position: 'fixed',
      top: '0',
      right: '0',
      bottom: '0',
      left: '0',
      backgroundColor: theme.palette.primary.dark,
    }),
    modal: ({ theme }: { theme: Theme }) => ({
      position: 'fixed',
      zIndex: theme.zIndex.modal,
      top: '50%',
      left: '50%',
      overflow: 'hidden',
      minWidth: 400,
      maxWidth: 600,
      boxSizing: 'border-box',
      padding: `${theme.offsets.sm}px ${theme.offsets.md}px`,
      backgroundColor: theme.palette.primary.light,
      borderRadius: theme.borderRadius.md,
      boxShadow: theme.shadows.xs,
      transform: 'translate(-50%, -50%)',
    }),
    header: ({ theme }: { theme: Theme }) => ({
      padding: theme.offsets.md,
      margin: `-${theme.offsets.sm}px -${theme.offsets.md}px 0`,
      backgroundColor: theme.palette.primary.main,
    }),
    title: ({ theme }: { theme: Theme }) => ({
      padding: '0',
      margin: '0',
      fontSize: theme.fontSize.xl,
      fontWeight: 600,
    }),
    content: ({ theme }: { theme: Theme }) => ({
      margin: theme.offsets.md,
    }),
    text: ({ theme }: { theme: Theme }) => ({
      padding: '0',
      margin: '0',
      fontSize: theme.fontSize.sm,
      fontWeight: 400,

      '&:not(:first-child)': {
        marginBottom: theme.offsets.xs,
      },
    }),
  },
  { name: 'Modal' }
);
