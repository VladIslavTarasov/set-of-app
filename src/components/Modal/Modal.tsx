import React from 'react';
import { createPortal } from 'react-dom';

import { useTranslation } from 'react-i18next';

import Button from 'components/Button';

import style from './Modal.module.scss';

export interface ModalProps {
  title?: string;
  description?: string | string[];
  onClose?: () => void;
}

const Modal: React.FC<ModalProps> = ({ title, description, onClose, children }) => {
  const { t } = useTranslation('buttons');

  return createPortal(
    <div className={style.container}>
      <div className={style.modal} role="dialog">
        {title && (
          <header className={style.header}>
            <h3 className={style.title}>{title}</h3>
          </header>
        )}

        {description && (
          <div className={style.content}>
            {Array.isArray(description) ? (
              description.map(item => (
                <p key={item} className={style.text}>
                  {item}
                </p>
              ))
            ) : (
              <p className={style.text}>{description}</p>
            )}
          </div>
        )}

        {onClose && (
          <Button onClick={onClose} type="button">
            {t('close')}
          </Button>
        )}

        {children}
      </div>
    </div>,
    document.body
  );
};

export default Modal;
