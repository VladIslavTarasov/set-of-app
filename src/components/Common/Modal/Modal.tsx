import React from 'react';

import { createPortal } from 'react-dom';
import { useTranslation } from 'react-i18next';

import Button from 'components/Common/Button';
import { useTheme } from 'theme/theme';

import { useStyles } from './Modal.styles';

export interface ModalProps {
  title?: string;
  description?: string | string[];
  onClose?: () => void;
}

const Modal: React.FC<ModalProps> = ({ title, description, onClose, children }) => {
  const theme = useTheme();
  const classes = useStyles({ theme });
  const { t } = useTranslation('buttons');

  return createPortal(
    <div data-testid="modal" className={classes.container}>
      <div className={classes.modal} role="dialog">
        {title && (
          <header className={classes.header}>
            <h3 className={classes.title}>{title}</h3>
          </header>
        )}

        {description && (
          <div className={classes.content}>
            {Array.isArray(description) ? (
              description.map(item => (
                <p key={item} className={classes.text}>
                  {item}
                </p>
              ))
            ) : (
              <p className={classes.text}>{description}</p>
            )}
          </div>
        )}

        {onClose && (
          <Button onClick={onClose} type="button">
            {t('close')}
          </Button>
        )}

        {children}
      </div>
    </div>,
    document.body
  );
};

export default Modal;
