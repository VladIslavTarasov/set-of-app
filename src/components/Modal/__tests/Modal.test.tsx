import React from 'react';

import { render } from '@testing-library/react';

import Modal from '..';

describe('components/Modal', () => {
  it('render with title', () => {
    const { container } = render(<Modal title="Title" />);
    expect(container).toMatchSnapshot();
  });

  it('render with description', () => {
    const { container, rerender } = render(<Modal description="description" />);
    expect(container.querySelectorAll('p.text').length).toBe(1);
    expect(container).toMatchSnapshot();

    rerender(<Modal description={['description1', 'description2', 'description3']} />);
    expect(container.querySelectorAll('p.text').length).toBe(3);
    expect(container).toMatchSnapshot();
  });

  it('render with button', () => {
    // eslint-disable-next-line
    const { container } = render(<Modal onClose={() => {}} />);
    expect(container).toMatchSnapshot();
  });

  it('render with children', () => {
    const { container } = render(
      <Modal>
        <h5>Title</h5>
        <p>Description</p>
        <button type="button">Close</button>
      </Modal>
    );
    expect(container).toMatchSnapshot();
  });
});
