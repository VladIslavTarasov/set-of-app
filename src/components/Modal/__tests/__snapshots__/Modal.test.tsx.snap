// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Modal render with button 1`] = `
<div
  class="container"
  data-testid="modal"
>
  <div
    class="modal"
    role="dialog"
  >
    <button
      class="button md"
      type="button"
    >
      close
    </button>
  </div>
</div>
`;

exports[`components/Modal render with children 1`] = `
<div
  class="container"
  data-testid="modal"
>
  <div
    class="modal"
    role="dialog"
  >
    <h5>
      Title
    </h5>
    <p>
      Description
    </p>
    <button
      type="button"
    >
      Close
    </button>
  </div>
</div>
`;

exports[`components/Modal render with description 1`] = `
<div
  class="container"
  data-testid="modal"
>
  <div
    class="modal"
    role="dialog"
  >
    <div
      class="content"
    >
      <p
        class="text"
      >
        description
      </p>
    </div>
  </div>
</div>
`;

exports[`components/Modal render with description 2`] = `
<div
  class="container"
  data-testid="modal"
>
  <div
    class="modal"
    role="dialog"
  >
    <div
      class="content"
    >
      <p
        class="text"
      >
        description1
      </p>
      <p
        class="text"
      >
        description2
      </p>
      <p
        class="text"
      >
        description3
      </p>
    </div>
  </div>
</div>
`;

exports[`components/Modal render with title 1`] = `
<div
  class="container"
  data-testid="modal"
>
  <div
    class="modal"
    role="dialog"
  >
    <header
      class="header"
    >
      <h3
        class="title"
      >
        Title
      </h3>
    </header>
  </div>
</div>
`;
